### PR TITLE
basic nix support, builds subkey and provides it to other packages

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
Basic nix support.
  
- What does it do? 

Builds subkey via nix

- What important points should reviewers know? 

`nix run .#subkey`, ensure flakes enabled (zero to nix installer is default)

Is there something left for follow-up PRs?
no follow up, so may build other tools if there are later

other usage of nix can be seen here https://github.com/paritytech/zombienet for example

### CODE REVIEW

**Comment by ggwpez** [comment]
Nix usage in the past had the problem that is was got out-dated since there was no CI to check that it still builds.  
Maybe a good follow-up would be to add that.

**Comment by ggwpez** [comment]
> @ggwpez i am not agains CI PR sure, but that could be out of scope for Parity to merge it now? so nixified subkey tool is nice regardless imho. you can onliner it it on nixos ;)

Yes sure it is out of scope.  

I personally dont use nix/flake, so cant approve. But maybe @bkchr can :smile:

**Review by bkchr — APPROVED** [review]

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+          rust-libs = {
+            buildInputs = with pkgs; [ openssl ] ++ darwin;
+            nativeBuildInputs = rust-native-build-inputs;
+            doCheck = false;
+          };
+          rust-deps = pkgs.makeRustPlatform {
+            inherit pkgs;
+            cargo = pkgs.rust-bin.beta.latest.default;
+            rustc = pkgs.rust-bin.beta.latest.default;
```
Comment (andresilva) [inline flake.nix:58]: stable works as well, and we could even use the rust toolchain from nixpkgs directly (that's I use to build subkey on my computer).

**Review by andresilva — CHANGES_REQUESTED** [review]
Thanks for the PR and really sorry for this not getting merged sooner. I started making some suggestions but they became too much to add here, so here's a commit with all my proposed changes: https://github.com/paritytech/substrate/commit/77be00ab11642c412c6eaa26e54b4ca64cc4f0c3. Some are just style nits, but some other changes make the definition simpler (like using `bindgenHook` instead of manually setting up clang), feel free to include whatever you think makes sense.

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
@dzmitry-lahoda could you fix the conflicts so that this can be merged? I can't push to your branch.

**Pull request closed without merging**